synapsemonitor/monitor.py

```diff
@@ -113,6 +113,47 @@ def _find_modified_entities_file(syn: Synapse, syn_id: str, days: int = 1) -> li
     return []
 
 
+def _traverse(
+    syn: Synapse,
+    synid_root: str,
+    include_types: typing.List = ["folder", "file", "project"],
+) -> list:
+    """Traverse Synapse entity hierarchy to gather all descendant
+    entities of a root entity.
+    Args:
+        syn: Synapse connection
+        synid_root: Synapse ID of root entity.
+        include_types: Must be a list of entity types (ie. [“folder”,”file”])
+            which can be found here:
+            http://docs.synapse.org/rest/org/sagebionetworks/repo/model/EntityType.html
+    Returns:
+        List of descendant Synapse IDs and root Synapse ID
+    """
+
+    synid_desc = []
+
+    # full traverse depends on examining folder entities, even if not requested
+    include_types_mod = include_types.copy()
+    if "folder" not in include_types_mod:
+        include_types_mod.append("folder")
+
+    synid_children = syn.getChildren(parent=synid_root, includeTypes=include_types_mod)
+    for synid_child in synid_children:
+        synid_desc = synid_desc + _traverse(
+            syn=syn, synid_root=synid_child["id"], include_types=include_types
+        )
+
+    # only return folder entities if requested
+    entity = syn.get(synid_root, downloadFile=False)
```

Comment:
This chunk of code is a bit confusing. what is it is a project?  What is the motivation here?

Reply:
Rewrote the boolean logic statement to hopefully make it clearer.  Now the boolean reads "if the user requested 'folder' entities OR if the entity type is not a folder, return the entity as a descendant"

This statement ensures that if the user only requests non-folder entities then only non-folder entities are returned as descendants in the traverse.  All other entity types can be filtered directly through the getChildren(includeTypes) parameter but folders are different.  This is because we need to traverse folder entities in case they contain child entities of the requested type (e.g. child folder of the root contains files).  But  at the end of each traverse step, we do not want to return folder entities if the user didn't request them.  

I was assuming "Project" entity types can only be root synapse IDs, which would work fine as implemented, but is this incorrect?  Specifically, would a project entity ID ever be returned through the getChildren() function?

Reply:
That is true but when you pass in a project ID as the first synapse id, that would be included. It's interesting because here you are saying you can pass in a folder id but you don't include any folder IDs in results but when you pass in a project, it will always return the project Id. (I think?)

Reply:
oh, I see.  you're right.  I'll make the change

Reply:
Maps entity type from concreteType and getChildren includeTypes and then checks that the entity_type is in the requested types before returning.  Mapping only known to work for file, folder, and project for now.